### PR TITLE
Refactor to pave way for LSO tracking

### PR DESCRIPTION
**Pull request by hachikuji**

### CODE REVIEW

File: core/src/main/scala/kafka/log/ProducerIdMapping.scala [hunk 1]
```diff
@@ -30,16 +30,70 @@ import org.apache.kafka.common.utils.{ByteUtils, Crc32}
...
+    this.lastOffset = lastOffset
   }
+
+  def append(batch: RecordBatch): Unit =
+    append(batch.producerEpoch, batch.baseSequence, batch.lastSequence, batch.maxTimestamp, batch.lastOffset)
+
+  def append(entry: ProducerIdEntry): Unit =
+    append(entry.epoch, entry.firstSeq, entry.lastSeq, entry.timestamp, entry.lastOffset)
+
+  def toSnapshotEntry: ProducerIdEntry =
```
Comment (apurvam) [inline core/src/main/scala/kafka/log/ProducerIdMapping.scala]: Maybe name this `lastEntry`?

**Comment by guozhangwang** [comment]
@hachikuji could you describe the major refactoring changes for this PR?

**Comment by hachikuji** [comment]
@guozhangwang Sure. The major objective was providing the hooks for `ProducerIdMapping` to extract more information out of the appended records. For LSO tracking, I need to see when transactions begin and end for each producer. With `ProducerAppendInfo`, we can now see all of the individual record updates and I can update the transaction state accordingly. As I was making this change, I realized that we could consolidate some of the validation that we were doing in `Log` and simplify the logic a bit. For example, instead of catching duplicate exceptions and then figuring out how to handle it, I directly checked for the one interesting case during the single validation pass in `Log.analyzeAndValidate`.

**Comment by hachikuji** [comment]
Tests are passing locally after rebasing. Going to merge shortly.

**Pull request merged**